pytorch_lightning/utilities/jsonargparse_utils.py

```diff
@@ -0,0 +1,150 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from typing import Type
+from jsonargparse import ArgumentParser, ActionConfigFile
+from pytorch_lightning import Trainer, LightningModule, LightningDataModule
+from pytorch_lightning.callbacks import Callback
+
+
+class LightningArgumentParser(ArgumentParser):
+    def __init__(self, *args, **kwargs):
+        """Initialize argument parser that supports configuration file input"""
+        super().__init__(*args, **kwargs)
+        self.add_argument('--cfg',
+            action=ActionConfigFile,
+            help='Path to a configuration file in json or yaml format.')
+
+
+    def add_trainer_args(
+        self,
+        trainer_class: Type[Trainer] = Trainer,
+        nested_key: str = 'trainer'
+    ):
+        """
+        Adds arguments from a trainer class to a nested key of the parser
+
+        Args:
+            trainer_class: Optional extension of the Trainer class.
+            nested_key: Name of the nested namespace where parsed arguments are stored.
+        """
+        assert issubclass(trainer_class, Trainer)
+        self.add_class_arguments(trainer_class, nested_key)
+
+
+    def add_module_args(
+        self,
+        module_class: Type[LightningModule],
+        nested_key: str = 'module'
+    ):
+        """
+        Adds arguments from a module class to a nested key of the parser
+
+        Args:
+            module_class: A LightningModule class.
+            nested_key: Name of the nested namespace where parsed arguments are stored.
+        """
+        assert issubclass(module_class, LightningModule)
+        self.add_class_arguments(module_class, nested_key)
+
+
+    def add_datamodule_args(
+        self,
+        datamodule_class: Type[LightningDataModule],
+        nested_key: str = 'data'
+    ):
+        """
+        Adds arguments from a datamodule class to a nested key of the parser
+
+        Args:
+            datamodule_class: A LightningDataModule class.
+            nested_key: Name of the nested namespace where parsed arguments are stored.
+        """
+        assert issubclass(datamodule_class, LightningDataModule)
+        self.add_class_arguments(datamodule_class, nested_key)
+
+
+class SaveConfigCallback(Callback):
+    """Callback that saves a trainer_cli config to the log_dir when training starts"""
+
+    def __init__(self, parser, config):
+        self.config_dump = parser.dump(config, skip_none=False)
+
+
+    def on_train_start(self, trainer, pl_module):
+        config_path = os.path.join(trainer.logger.log_dir, 'config.yaml')
+        with open(config_path, 'w') as outstream:
+            outstream.write(self.config_dump)
+
+
+def trainer_cli(
+    model_class: Type[LightningModule],
+    datamodule_class: Type[LightningDataModule] = None,
+    save_config_callback: Type[Callback] = SaveConfigCallback,
```

Comment:
well if there will be some derived Saver which would take also init arguments it won't work, consider passing instances instead...

Reply:
The SaveConfigCallback class requires to be instantiated with the parser and parsed config which are not yet available outside of trainer_cli. It could be done differently, but I guess it would require a more tight integration. Would need to get some ideas about this and what would be the benefits.

Reply:
so it means that any custom Saver can't have any init arguments? some each change has to be made with extra child?

Reply:
The idea I had here was that people would extend it if for example additionally to saving it in the log directory they would want to log it as an artifact. For this there is no need for new init arguments. What would be an example in which more init arguments are required? Anyway if someone wants something different they could add a custom callback like how @SeanNaren proposes in https://github.com/PyTorchLightning/pytorch-lightning/pull/4492#issuecomment-721194868